src/main/kotlin/uk/gov/justice/digital/hmpps/approvedpremisesapi/service/DomainEventService.kt

```diff
@@ -207,35 +213,130 @@ class DomainEventService(
       ),
     )
 
-    if (emitDomainEventsEnabled) {
-      val snsEvent = SnsEvent(
-        eventType = typeName,
-        version = 1,
-        description = typeDescription,
-        detailUrl = detailUrl,
-        occurredAt = domainEvent.occurredAt.atOffset(ZoneOffset.UTC),
-        additionalInformation = SnsEventAdditionalInformation(
-          applicationId = domainEvent.applicationId,
+    if (!emitDomainEventsEnabled) {
+      log.info("Not emitting SNS event for domain event because domain-events.cas1.emit-enabled is not enabled")
+      return
+    }
+
+    if (asyncSaveEnabled) {
+      doAsyncSave(
+        domainEvent,
+        typeName,
+        typeDescription,
+        detailUrl,
+        crn,
+        nomsNumber,
+      )
+    } else {
+      doSyncSave(
+        domainEvent,
+        typeName,
+        typeDescription,
+        detailUrl,
+        crn,
+        nomsNumber,
+      )
+    }
+  }
+
+  final fun retryTemplate(): RetryTemplate {
+    val retryTemplate = RetryTemplate()
+    val backOffPolicy = FixedBackOffPolicy()
+    backOffPolicy.backOffPeriod = BACK_OFF_PERIOD
+    retryTemplate.setBackOffPolicy(backOffPolicy)
+    val retryPolicy = SimpleRetryPolicy()
+    retryPolicy.maxAttempts = MAX_ATTEMPTS_RETRY
+    retryTemplate.setRetryPolicy(retryPolicy)
+    return retryTemplate
+  }
+
+  fun doAsyncSave(
+    domainEvent: DomainEvent<*>,
+    typeName: String,
+    typeDescription: String,
+    detailUrl: String,
+    crn: String,
+    nomsNumber: String,
+  ) = runBlocking {
+    val snsEvent = SnsEvent(
```

Comment:
Can we add shared functions to generate the `snsEvent` and  `publishRequest`, as they're essentially the same in `doAsyncSave` and `doSyncSave`?